### PR TITLE
flyctl: add installCheckPhase to test binary successfully runs after build

### PR DESCRIPTION
**Pull request by techknowlogick**
###### Description of changes

followup on https://github.com/NixOS/nixpkgs/pull/172247 to add test to ensure binary runs after build

###### Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [x] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.05 Release Notes (or backporting 21.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2205-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/development/web/flyctl/default.nix [hunk 1]
```diff
@@ -35,6 +35,12 @@ buildGoModule rec {
     go test ./... -ldflags="-X 'github.com/superfly/flyctl/internal/buildinfo.buildDate=1970-01-01T00:00:00+0000'"
   '';
 
+  doInstallCheck = true;
+
+  installCheckPhase = ''
+    $out/bin/flyctl version | grep ${version} > /dev/null
```
Comment (davidak) [inline pkgs/development/web/flyctl/default.nix:41]: hi @techknowlogick 

we now have `testVersion` for this

https://nixos.org/manual/nixpkgs/unstable/#chap-testers

Comment (techknowlogick) [reply pkgs/development/web/flyctl/default.nix:45]: thanks for feedback :) I've updated PR to switch to that.

**Comment by davidak** [comment]
Result of `nixpkgs-review pr 172428` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>flyctl</li>
  </ul>
</details>

File: pkgs/development/web/flyctl/default.nix [hunk 2]
```diff
@@ -32,9 +32,13 @@ buildGoModule rec {
   '';
 
   postCheck = ''
-    go test ./... -ldflags="-X 'github.com/superfly/flyctl/internal/buildinfo.buildDate=1970-01-01T00:00:00+0000'"
+    go test ./... -ldflags="-X 'github.com/superfly/flyctl/internal/buildinfo.buildDate=1970-01-01T00:00:00Z'"
   '';
 
+  passthru.tests.version = testers.testVersion {
+    command = "flyctl version";
```
Comment (davidak) [inline pkgs/development/web/flyctl/default.nix:39]:
```suggestion
    package = flyctl;
    command = "flyctl version";
```

it needs the package

```
[nix-shell:~/.cache/nixpkgs-review/pr-172428]$ nix-build nixpkgs/ -A flyctl.tests
error: 'testVersion' at /home/davidak/.cache/nixpkgs-review/pr-172428/nixpkgs/pkgs/build-support/testers/default.nix:7:5 called without required argument 'package'

       at /home/davidak/.cache/nixpkgs-review/pr-172428/nixpkgs/pkgs/development/web/flyctl/default.nix:38:28:

           37|
           38|   passthru.tests.version = testers.testVersion {
             |                            ^
           39|     command = "flyctl version";
```

**Comment by davidak** [comment]
Result of `nixpkgs-review pr 172428` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>flyctl</li>
  </ul>
</details>

**Comment by techknowlogick** [comment]
@davidak its failing using passthrough because read only homedir. Im unfamiliar with the newer way of doing the check, does it have any hooks for being able to export HOME env var with mktemp -d? I couldn't find anything in docs.

**Comment by davidak** [comment]
Yes, i was looking into this too:

```
[nix-shell:~/.cache/nixpkgs-review/pr-172428-1]$ nix-build nixpkgs/ -A flyctl.tests
this derivation will be built:
  /nix/store/0qxn9fqpb1755idvwb3y3q862wyrv6z0-flyctl-0.0.323-test-version.drv
building '/nix/store/0qxn9fqpb1755idvwb3y3q862wyrv6z0-flyctl-0.0.323-test-version.drv'...
Error accessing config directory at $HOME/.fly mkdir /homeless-shelter: permission denied
Error failed creating config directory: mkdir /homeless-shelter: permission denied
error: builder for '/nix/store/0qxn9fqpb1755idvwb3y3q862wyrv6z0-flyctl-0.0.323-test-version.drv' failed with exit code 1;
```

I don't know how to fix it right now.

I also found setting HOME to temp dir...

https://github.com/NixOS/nixpkgs/issues/16144#issuecomment-225422439
https://githubmemory.com/repo/svanderburg/node2nix/issues/217

`export HOME=$(mktemp -d)`

There seem to be no prePhase.

Any idea how to fix this @Artturin?

**Comment by Artturin** [comment]
Set the command arg ` https://nixos.org/manual/nixpkgs/unstable/#tester-testVersion

**Comment by davidak** [comment]
@Artturin do you mean it can be multiline? then we should also add an example for that

**Comment by Artturin** [comment]
`VAR=value thecommand --version`

**Comment by davidak** [comment]
Result of `nixpkgs-review pr 172428` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>flyctl</li>
  </ul>
</details>

**Comment by davidak** [comment]
That seem to work, but still fails. No useful error message:

```
[nix-shell:~/.cache/nixpkgs-review/pr-172428-2]$ nix-build nixpkgs/ -A flyctl.tests
this derivation will be built:
  /nix/store/x9smfa2spcdz0ha00b2pwnk7vmqqgm5k-flyctl-0.0.323-test-version.drv
building '/nix/store/x9smfa2spcdz0ha00b2pwnk7vmqqgm5k-flyctl-0.0.323-test-version.drv'...
error: builder for '/nix/store/x9smfa2spcdz0ha00b2pwnk7vmqqgm5k-flyctl-0.0.323-test-version.drv' failed with exit code 1
```

Maybe because it contains more than just the version?

```
[nix-shell:~/.cache/nixpkgs-review/pr-172428-2]$ flyctl version
flyctl v0.0.323 linux/amd64 Commit: v0.0.323 BuildDate: 1970-01-01T00:00:00Z
```

**Comment by techknowlogick** [comment]
@Artturin @davidak awesome. that worked. Thanks for your guidance :)

**Comment by davidak** [comment]
```
[nix-shell:~/.cache/nixpkgs-review/pr-172428-2]$ grep -Fw "0.0.323" - <<< "flyctl v0.0.323 linux/amd64 Commit: v0.0.323 BuildDate: 1970-01-01T00:00:00Z"

[nix-shell:~/.cache/nixpkgs-review/pr-172428-2]$ echo $?
1

[nix-shell:~/.cache/nixpkgs-review/pr-172428-2]$ grep -Fw "v0.0.323" - <<< "flyctl v0.0.323 linux/amd64 Commit: v0.0.323 BuildDate: 1970-01-01T00:00:00Z"
flyctl v0.0.323 linux/amd64 Commit: v0.0.323 BuildDate: 1970-01-01T00:00:00Z

[nix-shell:~/.cache/nixpkgs-review/pr-172428-2]$ echo $?
0
```

File: pkgs/development/web/flyctl/default.nix [hunk 3]
```diff
@@ -32,9 +32,14 @@ buildGoModule rec {
   '';
 
   postCheck = ''
-    go test ./... -ldflags="-X 'github.com/superfly/flyctl/internal/buildinfo.buildDate=1970-01-01T00:00:00+0000'"
+    go test ./... -ldflags="-X 'github.com/superfly/flyctl/internal/buildinfo.buildDate=1970-01-01T00:00:00Z'"
   '';
 
+  passthru.tests.version = testers.testVersion {
+    package = flyctl;
```
Comment (davidak) [inline pkgs/development/web/flyctl/default.nix:39]:
```suggestion
    package = flyctl;
    version = "v${flyctl.version}";
```
does it work like this?

Comment (davidak) [reply pkgs/development/web/flyctl/default.nix:39]: that works

```
[nix-shell:~/.cache/nixpkgs-review/pr-172428-2]$ nix-build nixpkgs/ -A flyctl.tests
this derivation will be built:
  /nix/store/l5sbn1jyb88pqgzmnpj960kxy2z9hw2g-flyctl-0.0.323-test-version.drv
building '/nix/store/l5sbn1jyb88pqgzmnpj960kxy2z9hw2g-flyctl-0.0.323-test-version.drv'...
flyctl v0.0.323 linux/amd64 Commit: v0.0.323 BuildDate: 1970-01-01T00:00:00Z
/nix/store/knwy3jq2d906dbp5mlc36x3qyj45ai4y-flyctl-0.0.323-test-version
```

`version` is not known in `passthru.tests` context from the package

Comment (davidak) [reply pkgs/development/web/flyctl/default.nix:39]: maybe i'm wrong. it works also with `version`

@Artturin  what is more correct?

**Comment by davidak** [comment]
Result of `nixpkgs-review pr 172428` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>flyctl</li>
  </ul>
</details>

**Comment by davidak** [comment]
Result of `nixpkgs-review pr 172428` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>flyctl</li>
  </ul>
</details>

**Review by davidak — APPROVED** [review]

**Comment by davidak** [comment]
Thanks for your effort and patience. This was a complex edge-case!

**Pull request merged**